src/graph/sampling/randomwalks/get_node_types_gpu.cu

```diff
@@ -0,0 +1,90 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file graph/sampling/get_node_types_gpu.cu
+ * \brief DGL sampler
+ */
+
+#include <dgl/array.h>
+#include <dgl/base_heterograph.h>
+#include <dgl/runtime/device_api.h>
+#include <cuda_runtime.h>
+#include <utility>
+#include "randomwalks_impl.h"
+
+namespace dgl {
+
+using namespace dgl::runtime;
+using namespace dgl::aten;
+
+namespace sampling {
+
+namespace impl {
+
+template<DLDeviceType XPU, typename IdxType>
+TypeArray GetNodeTypesFromMetapath(
+    const HeteroGraphPtr hg,
+    const TypeArray metapath) {
+
+  uint64_t num_etypes = metapath->shape[0];
+  TypeArray result = TypeArray::Empty(
+      {metapath->shape[0] + 1}, metapath->dtype, metapath->ctx);
+  IdxType *result_data = static_cast<IdxType *>(result->data);
+
+  auto metapath_ctx = metapath->ctx;
+  auto cpu_ctx = DGLContext{kDLCPU, 0};
+  auto metapath_device = DeviceAPI::Get(metapath_ctx);
+  auto cpu_device = DeviceAPI::Get(cpu_ctx);
+  // use default stream
+  cudaStream_t stream = 0;
+
+  auto h_metapath_data = static_cast<IdxType*>(
+      cpu_device->AllocWorkspace(cpu_ctx, sizeof(IdxType) * (num_etypes)));
+  auto h_result_data = static_cast<IdxType*>(
+      cpu_device->AllocWorkspace(cpu_ctx, sizeof(IdxType) * (num_etypes + 1)));
+
+  metapath_device->CopyDataFromTo(static_cast<const IdxType*>(metapath->data), 0,
+                                  h_metapath_data, 0,
+                                  sizeof(IdxType) * (num_etypes),
+                                  metapath_ctx, cpu_ctx,
+                                  metapath->dtype, stream);
```

Comment:
There is an `NDArray::CopyTo(DLContext)` method esp. for default stream copies.

Also, it might be easier if you restrict the metapath array on CPU.

Reply:
If we are to allow this to happen on a non-default stream, we need a `metapath_device->StreamSync(metapath_ctx, stream);` before accessing the data on the CPU.